forms/upload.go

```diff
@@ -0,0 +1,80 @@
+package forms
+
+import (
+	"mime/multipart"
+
+	"go-google-scraper-challenge/helpers"
+	"go-google-scraper-challenge/models"
+
+	"github.com/beego/beego/v2/core/logs"
+	"github.com/beego/beego/v2/core/validation"
+)
+
+type UploadForm struct {
+	File       multipart.File
+	FileHeader *multipart.FileHeader
+	User       *models.User          `valid:"Required"`
+	Keywords   []string
+}
+
+// Valid adds custom validation to upload form, sets error when the validation failed.
+func (uf *UploadForm) Valid(v *validation.Validation) {
```

Comment:
This function has 4 level nested blocks. I  think it can flatten each validation functionality wrap into a separate function (file emptiness will check a function, corrupted file will check in a separate function). I did similar to make the `Valid` function [lighter](https://github.com/junan/google-scraper/blob/develop/forms/search_form.go#L34-L42)

Reply:
@junan How about adding `return` for the first 3 validation 🤔  IMO this way it's more simple and readable

```Go
        if uf.File == nil || uf.FileHeader == nil {
		err := v.SetError("File", "File cannot be empty")
		if err == nil {
			logs.Info("Failed to set error on validation")
		}

		return
	}

	fileType := helpers.GetFileType(uf.FileHeader)
	if fileType != "text/csv" {
		err := v.SetError("File", "Incorrect file type")
		if err == nil {
			logs.Info("Failed to set error on validation")
		}

		return
	}

	keywords, err := helpers.GetFileContent(uf.File)
	if err != nil {
		err := v.SetError("File", "Unreadable file")
		if err == nil {
			logs.Info("Failed to set error on validation")
		}

		return
	}

	if len(keywords) < 1 {
		err := v.SetError("File", "File should contains at least one keyword")
		if err == nil {
			logs.Info("Failed to set error on validation")
		}
	} else if len(keywords) > 1000 {
		err := v.SetError("File", "File contains too many keywords")
		if err == nil {
			logs.Info("Failed to set error on validation")
		}
	} else {
		uf.Keywords = keywords
	}
```

Reply:
Early returning will make it flatten and it will be better than deep nesting. But I think still this `Valid` function will be still big and if we add more validation criteria in the future(ex: validate certain keyword exist, file size(1MB, max like so) validation) it will be even bigger and any new developer need to read the details block to understand what CSV validation it does. But If we use warp each validation functionality into a separate function and use early return together, then the `Valid` function will be smaller and any developer will know what validations does it does just seeing the function name without looking into details implementation.

For example, like the bellow pseudo code
```go
func (uf *UploadForm) Valid(v *validation.Validation) {
....
    err := validateFilePresence(uf)
    if err != nil {
        return
    }
    err = validateFileType(uf)
     if err != nil {
         return
    }
...
}
``` 

What do you think?

Reply:
This is what I want to comment too. I agree with @junan to early returning combine with separate validation step to the new function to make the this function concise and simple.

Is it something like this @junan?
https://github.com/gutakk/go-google-scraper/blob/master/controllers/keyword.go#L75-L107

Reply:
@junan @gutakk Does this refactoring works for you guys? [3a9582e](https://github.com/carryall/go-google-scraper-challenge/pull/56/commits/3a9582ec48c11c7b06de86797c1307e024373c67)

Reply:
Looks great for me. This is optional, I think we can combine `validateKeywordsNotEmpty` and `validateKeywordsCountNotExceed` together and the error would be something like "CSV file must contain between 1 to 1000 keywords". This will allow user to understand what is the range of the csv content should be.

What do you think?

Reply:
updated in [60b0bb8](https://github.com/carryall/go-google-scraper-challenge/pull/56/commits/60b0bb89b05ec7173695c066301fbcaf451c5e8d)

Reply:
Looks good to me.

As Gut said above, `validateKeywordsNotEmpty` and `validateKeywordsCountNotExceed` can be combined into a single function